best-fit allocator

damiendoligez opened the pull request:
Add a best-fit allocator, which helps a lot with fragmentation, according to some testing done at Jane Street on memory-hungry programs.

Notes for the reviewer:

The main change is `freelist.c`. It is now has three main sections:
- next-fit allocator
- first-fit allocator
- best-fit allocator

The first two are the old allocator, split into two separate allocators. Reading these sections should be done in parallel with two copies of the old code (git only does the diff with the first, of course). The third is entirely new code.

I had to change the API of `caml_fl_merge_block` to avoid the cost of repeated coalescing in the best-fit case.

damiendoligez committed:
best-fit allocator

damiendoligez committed:
add Changes entry

alainfrisch commented:
That sounds great! Do you have examples of small programs that exhibit significant gains, or otherwise some intuition about allocation schemes likely to benefit from the new policy?  Does the reduced fragmentation comes with some higher GC cost?  Do you expect that the new policy could become the default at some point, or is the overhead possibly too large in some common cases?

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+  value free;
+  value *merge;
+} bf_small_fl [BF_NUM_SMALL + 1];
+int bf_small_map = 0;
```
```suggestion
static int bf_small_map = 0;
```

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+   with a splay tree (Sleator & Tarjan).
+*/
+
+/* [BF_NUM_SMALL] must be at least 4 for this code to work
```
It seems like it needs to be at least 5, to fit a `struct large_free_block`?

damiendoligez replied:
Small block sizes start at 1, so `BF_NUM_SMALL` is also the size of the largest small block. If it is 4, then the smallest large block is 5, enough for a `struct large_free_block`.

stedolan replied:
Got it, thanks.

bluddy commented:
Pretty sure we're trading performance for space -- the GC cost must be higher. It'd be nice to see a comparison.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+   White if it is a fragment (wosize = 0)
+   Black if it is a small block (0 < wosize <= BF_NUM_SMALL)
+   Blue if it is a large block (BF_NUM_SMALL < wosize)
+*/
```
The comment and code here are inconsistent about whether a "fragment" is a zero-word block left over after allocation, unsuitable for future allocation, or a <= BF_NUM_SMALL-word block which may be used for a small future allocation.

damiendoligez replied:
You're right, I'm using "fragment" for two different things. Renamed with some help from synonym.com.

lpw25 commented:
I'll give some more practical numbers in a bit, but since you asked for small pathological examples, here's one:

```ocaml
let r = ref []
let r2 = ref [||]

let f () =
  (* Allocate 1 gb of mem, allocate some more and compact to make
     sure we have 1gb of mem + 80% gc overhead allocated right now. *)
  let len = 1024 * 1024 * 1024 / (Sys.word_size / 8) in
  let gb_of_mem = Array.init len (fun i -> Some i) in
  ignore (Array.copy gb_of_mem : _ array);
  Gc.compact ();
  let count = ref 0 in
  Printf.printf "initialized\n%!";
  while true; do
    (* And now: allocate continuously. Every 250 words of transiently allocation on
       average, keep one word alive in [gb_of_mem] to create holes in any gap.  We do not
       increase live memory though since we replace some other allocation from [r]. *)
    let conses = Random.int 500 / 3 in
    r2 := [||];
    for _ = 0 to conses - 1; do r := 1 :: !r; done;
    gb_of_mem.(Random.int len) <- Some !count;
    (Sys.opaque_identity r) := [];
    if !count mod 1500000 = 0 then
      let live_words = len * 3 (* 1 for the array + 2 for the Some *) in
      let stat = Gc.stat () in
      Printf.printf "chunks=%d, largest free: %.2fMW, heap_words=%.2fGW, live_fraction=%f%%\n%!"
        stat.heap_chunks
        (Float.of_int stat.largest_free *. 1e-6)
        (Float.of_int stat.heap_words *. 1e-9)
        (100. *. Float.of_int live_words /. Float.of_int stat.heap_words)
    else if !count mod 10000 = 5000 then
      (* from times to times, the program needs to do a big allocation, and because
         of all the fragmentation, the gc grows the heap *)
      r2 := Array.make 50_000 0;
    incr count;
  done;
;;

let () = f ()
```
I think this should grow its heap steadily until it eventually compacts with next fit, even though the number of live words is constant.  With best fit it will stop growing the heap once the overhead matches the given `space_overhead` parameter.

I'm not sure that these small examples are the best way to evaluate these things, but you did ask. Judging from my results with larger programs, there are other pathological cases fixed by best-fit which don't look like this, but I don't have small examples of those to hand.

lpw25 commented:
(The above example is due to @sliquister)

alainfrisch commented:
Thanks @lpw25 . I think this is useful to get an intuition on when this new policy can help.  For such a program, I understand that fragmentation is better, and the heap will stop growing earlier than with other strategies.  Do you have some idea of the runtime overhead due the new policy (as long as the heap remains reasonnable with former policies)?

lpw25 commented:
Mostly I haven't seen an observable overhead.

A couple of programs had some small overhead (1-2%) which seemed to just be a product of having a smaller heap. Such overheads were easily compensated for by just relaxing `space_overhead` a bit -- which you can do since you are using less heap now.

I think another way to look at the above is that the current GC heuristics for ensuring that `space_overhead` is correct assume that there is no fragmentation. Since there is fragmentation we are actually getting more overhead than `space_overhead` usually -- so a setting of 80% is secretly more like a setting of 100%. When best-fit reduces fragmentation it makes the `space_overhead` heuristics more accurate -- so to keep equivalent behaviour you should increase the value of the `space_overhead` parameter.

In some cases there have been notable improvements in execution times. One real program had a more than 50% reduction in execution time with best-fit because it was hitting a pathological case in the next-fit allocator that lead to it spending 65% of its time traversing the free list.

stedolan commented:
I don't have time this evening, but I'll read this properly next week. It looks great - the splay tree of lists gives a much more efficient search than the current allocator. I haven't read the code in detail yet, but I'm curious about the design and allocation policy for small allocations. (For large allocations, the policy seems very simple - precise best-fit - and the implementation is really nice).

The preallocation mechanism, if I understand it correctly, allocates space for 100 small objects when the first allocation is requested so that the next 99 allocations are fast. This happens if there is no currently available small slot of the correct size or greater. So, if I allocate 500 cons cells, I only do 5 accesses to the splay tree.

However, if I allocate a single 16-word object before my 1000 cons cells, then I preallocate 1700 words of space and none of my 500 subsequent cons-cell allocations hit the preallocation fast-path, all hitting the `bf_split_small`/`bf_insert_fragment_small` path instead. I'm not sure how much slower this is, but it seems a bit unfortunate.

As an alternative, might it be better to avoid ever splitting small blocks, and only merging them if the result of the merge is bigger than `BF_NUM_SMALL`? That way, we'd hit the fast preallocated path in this case as well. (I imagine there's some horrible pathological case here. It might be that the current policy is better, but this is the part of the design that I understand least).

Also, have you put any thought into how this design might work with multicore?

bluddy commented:
I guess that makes sense. If you have fragmentation, not only are you traversing the free list for longer, you're also accessing fresh memory more often, which could reduce cache efficiency, which would dominate other performance considerations.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+/* Search for the least node that is large enough to accomodate the given
+   size. Return in [next_lower] an upper bound on the size of the
+   next-lower node in the tree, or BF_NUM_SMALL if there is no such node.
+*/
```
If the smallest node in the tree has size 42, then `bf_search_best(42, &lower)` will return 42 in lower, not BF_NUM_SMALL as the comment suggests. (This is probably fine, in which case only the comment needs changing)

damiendoligez replied:
Fixed the comment.

lpw25 commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1384 @@ header_t *caml_fl_merge_block (value bp)
+        }
+        Next_small (Val_hp (p)) = Val_NULL;
+        /* Put the chain in the free list. */
+        bf_small_fl[wosz].free = Next_small (Val_hp (big));
```
This assumes that `bf_small_fl[wosz]` is currently empty. However, the call to `bf_allocate_from_tree` above on line 1533 could have added a node to it. This leaks a node and results in an assertion failure with the debug runtime.

damiendoligez replied:
Good catch indeed!

Do you have a simple repro case? I have a fix but I'd like to test it.

lpw25 replied:
Afraid not. I think it is causing a segfault in [this benchmark](https://github.com/OCamlPro/ocaml-benchs/blob/master/chameneos-redux/chameneos_redux_evtchn.ml) about 1 time in 20. I'm just checking at the moment that this is indeed the cause.

damiendoligez commented:
A few answers before I go on vacation:

@alainfrisch

> some intuition about allocation schemes likely to benefit from the new policy?

Our experience so far seems to show that all programs with large heaps will benefit to some extent. In normal cases, because the program fits in a noticeably smaller heap; in pathological cases because the compaction will trigger much less often.

> Does the reduced fragmentation comes with some higher GC cost?

The cost of maintaining and using the best-fit data structures doesn't seem to be higher than the simple next-fit free-list. But the space-time trade-off of the GC is actually controlled by the GC speed feedback loop, and when your heap is smaller (for example because of reduced fragmentation) it will spend more time in the GC. You can then recover your previous speed by increasing the space_overhead parameter of the GC.

> Do you expect that the new policy could become the default at some point, or is the overhead possibly too large in some common cases?

Yes, I expect to switch the default to best-fit, as the overhead is (as far as we can tell) negligible.

@stedolan

>However, if I allocate a single 16-word object before my 1000 cons cells, then I preallocate 1700 words of space and none of my 500 subsequent cons-cell allocations hit the preallocation fast-path, all hitting the bf_split_small/bf_insert_fragment_small path instead. I'm not sure how much slower this is, but it seems a bit unfortunate.

We could partly alleviate this problem by pre-allocating a fixed number of words (instead of blocks) so the list of 16-word blocks would be much shorter.

>As an alternative, might it be better to avoid ever splitting small blocks, and only merging them if the result of the merge is bigger than BF_NUM_SMALL? That way, we'd hit the fast preallocated path in this case as well. (I imagine there's some horrible pathological case here. It might be that the current policy is better, but this is the part of the design that I understand least).

I'm really wary of deviating too much from a strict best-fit policy because it's hard to tell what will happen in real programs. For example Wilson et al[1] observes that many programs have multiple phases during which they allocate some set of sizes, so it's expected that a program will allocate lots of 16-word blocks for some time, then suddenly switch to 2-word blocks. If you never split small blocks, you'll be stuck with a long list of 16-word blocks.

>Also, have you put any thought into how this design might work with multicore?

The small lists can be per-domain, and the splay tree will need locking. I'm worried about the contention on that lock. Maybe it can be reduced by splaying less often and using a single-writer/multiple-readers lock. If that doesn't work, we'll need to switch to a different data structure (and maybe a relaxed policy).

lpw25 commented:
> I'm worried about the contention on that lock. Maybe it can be reduced by splaying less often and using a single-writer/multiple-readers lock.

Note to our future selfs: there are approaches like "Lazy Splaying" which apparently reduce the contention in concurrent cases whilst still preserving good performance on unbalanced workloads, so we might want to look at them if this does turn out to be an issue.

damiendoligez committed:
fix a memory lead in bf_allocate

damiendoligez committed:
When the small free lists are empty, skip the tree search (if possible)
and allocate directly from the smallest block of the tree.

This removes the need for pre-loading the small free lists.

damiendoligez committed:
add a debug macro for readability

damiendoligez commented:
@lpw25 @stedolan This is the version with direct allocation in the smallest heap block. Also available for 4.08 at https://github.com/damiendoligez/ocaml/archive/4.08.1+best-fit.tar.gz

alainfrisch commented:
We have a program where repeatedly executing some action lead to memory growing each time significantly (to more than 1.5Gb) with next-fit but remains reasonnable (at around 500Mb) with first-fit; unfortunately, runtime is multiplied almost by 3 with first-fit.  Is that the kind of slowdown expected with first-fit, and is there some hope than the new policy would bring the best of both world?  (I suspect that fragmentation is due to the use of demarshaling -- I wonder whether demarshalling into small blocks wouldn't be better than the current strategy performance-wise in some cases; more GC-related overhead during demarshaling but less fragmentation...).

alainfrisch commented:
Answering to myself, with the help of @nojb who cherry-picked the new policy on our local version: the new policy looks terrific for our use case!

Some results running repeatedly some heavy calculation, monitoring the time taken in each run and the process memory usage after it, with the three policies:

|   |  policy 0 | policy 1 | policy 2 |
|---|-----------|----------|-----------|
| 1st run |  28s/855Mb | 47s/566Mb | 27s/545Mb
|  2nd run | 21s/1164Mb | 57s/584Mb | 22s/563Mb
| 3rd run | 21s/1316Mb | 59s/606Mb | 23s/572Mb


The new policy is only a tiny bit slower than policy 0 (perhaps in the noise), and even better than policy 1 in terms of memory usage.

With another payload where policy 1 was only slightly slower than policy 0 but clearly better in terms of memory usage, I observe a similar conclusion: policy 2 even better in terms of memory usage, and between 0 and 1 in terms of runtime.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+       collection never creates such runs.
+
+       This is a 2-state automaton that recognizes the regexp (w|wb)+
+       knowing that we start with a white block. */
```
Is this algorithm still necessary? My understanding was that such runs were only created by the small block preallocation mechanism, which is no longer present.

damiendoligez replied:
Simplified.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+/* [bf_large_least] is either NULL or a pointer to the smallest (leftmost)
+   block in the tree. In this latter case, the block must be alone in its
+   doubly-linked list (i.e. have [isnode] true and [prev] and [next] NULL)
+*/
```
I thought it would be difficult to maintain such a pointer through all the splaying operations, but by making it OK to set this pointer to NULL at any time you avoid all the difficult cases. Since the allocation code makes this pointer non-NULL at the next small allocation, the NULL state only lasts for a single small allocation. Nicely done!

damiendoligez replied:
The real trick is to have a pointer to the block itself. Splaying may move it around the tree but doesn't change its address.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+  unsigned long index;
+  unsigned char result;
+  result = _BitScanForward (&index, (unsigned long) x);
+  return (int) index + result;
```
I think this implementation is wrong.

  - You need to include `<intrin.h>` to get `_BitScanForward`
  - I think `_BitScanForward` numbers bits from 1 (like ffs), so adding result will get the wrong answer.
  - If x == 0, then `index` is unspecified after `_BitScanForward`

damiendoligez replied:
* fixed
* MS's specification doesn't say, but their example says `_BitScanForward(12)` returns `index=2`. https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanforward-bitscanforward64?view=vs-2019
* I could pre-set `index` to 0 and hope it isn't changed when x==0 but that's not really documented. Fixed.

All this shows that I need to do some precheck on this PR.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+/* How many blocks to pre-allocate at once for a small free list when
+   it and all the larger ones are empty.
+   100 should be a good number to amortize the overhead. */
+#define BF_PREALLOC_NUM 100
```
This is dead now.

stedolan commented:
This looks great! I have a couple of notes, but most of them are about comments.

lpw25 commented:
I've been running the latest version with the debug runtime on all our tests and I'm getting an assertion failure:

file gc_ctrl.c; line 168 ### Assertion failed: prev_hp == NULL || Color_hp (prev_hp) != Caml_blue || cur_hp == (header_t *) caml_gc_sweep_hp

which seems to indicate that there are fragments preceded by blue blocks on the heap, which I did not think was supposed to happen with the latest version.

stedolan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+static struct large_free_block *bf_large_least;
+/* [bf_large_least] is either NULL or a pointer to the smallest (leftmost)
+   block in the tree. In this latter case, the block must be alone in its
+   doubly-linked list (i.e. have [isnode] true and [prev] and [next] NULL)
```
I think `prev` and `next` point back to the list node when the list is a singleton, rather than being `NULL`.

damiendoligez committed:
simplify bf_merge_block

damiendoligez committed:
runtime/freelist.c: make bf_small_map a static variable

Co-Authored-By: Stephen Dolan <mu@netsoc.tcd.ie>

damiendoligez committed:
fixes suggested by @stedolan

stedolan commented on runtime/freelist.c:
```diff
@@ -919,12 +915,13 @@ static struct large_free_block *bf_large_least;
   unsigned char result;
   result = _BitScanForward (&index, (unsigned long) x);
-  return (int) index + result;
+  return result ? (int) index + 1 : 0;
```
Is the +1 needed? My impression from the docs was that `_BitScanForward` returns a 1-based number like ffs (but I found the docs quite hard to read)

damiendoligez replied:
I had this in another comment but GitHub seems to have eaten it:

MS's specification doesn't say, but their example says _BitScanForward(12) returns index=2. https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanforward-bitscanforward64?view=vs-2019

I'll run it through precheck anyway.

stedolan replied:
Right, but 12 is 1010 so index=2 is 1-based and doesn't need any +1 to match ffs?

lthls replied:
I haven't followed much of the discussion but I believe 12 is 1100 in binary. Hope that clarifies the discussion a bit.

stedolan replied:
Hah! I'll learn to count next time. Thanks!

damiendoligez commented:
@alainfrisch Policy 2 appears slower but that is probably because the heap size is smaller, so the feedback loop that pilots the GC slices will make it work harder (the free list is smaller, so the mutator eats a larger proportion of it at each slice, and we must do the corresponding proportion of GC).
I've been told (by @lpw25) that you can recover your speed by increasing slightly the `space_overhead` parameter of the GC from its default of 80. In this way, you should be as fast as policy 0 while the heap remains much smaller.

stedolan commented:
I found a fix for the issue @lpw25 hit above, [patch available here](https://github.com/stedolan/ocaml/commit/05c6324411e3f0a3c911f67c4889e51582fbf92d).

The issue is to do with the remnants created by splitting a small block which is larger than the requested allocation. Previously, those were unconditionally colored `Caml_black`, which may be changed to `Caml_white` by the sweeper.

However, if such a remnant happens to be preceded by a free block, the sweeper will not merge them as it will not notice that the remnant is also free block. This can result in two adjacent free blocks, or a free block being adjacent to a one-word white fragment (as in the assertion failure above).

This situation is quite difficult to trigger. If the remnant is created after sweeping has passed it, then this situation does not occur. If the remnant is created before sweeping begins, then this situation also does not occur: `bf_init_merge`, run at the start of sweeping, removes all small black remnants from the free list and colors them `Caml_white` so that the sweeper can pick them up and merge them. So, to trigger this bug, a small black remnant must be split after sweeping has begun but before sweeping has gone too far.

A C stub that triggers this situation [appears here](https://gist.github.com/stedolan/eecf5e21af170e1cd69d0bc388c3d5a5). Under the debug runtime with best-fit enabled by `OCAMLRUNPARAM=a=2`, this program crashes in `Gc.stat` without this patch and succeeds with this patch.

While the issue is very subtle, the fix is simple enough: when creating a small remnant, if sweeping has begun but not yet reached this point, the remnant is colored `Caml_white` so that the sweeper picks it up. Such white remnants are not added to the freelist.

This also tightens the invariants on the free list slightly: small blocks on the freelist are now always black or blue, never white. (The invariant checking in bf_check and elsewhere has been updated accordinly)

damiendoligez commented:
Wow, thanks for the awesome detective work! Your patch looks good, and I think we can optimize it further: now the sweeper will never see those black headers, so we can get rid of them and color everything white right away. Then we can move the test (whether to insert or leave to the sweeper) to the `insert_remnant`* functions.

damiendoligez committed:
fix for the problem found by @lpw25 and debugged by @stedolan (https://github.com/ocaml/ocaml/pull/8809#issuecomment-531861299)

damiendoligez commented:
@stedolan If you could review this latest commit, it's the same idea as yours but it touches different parts of the code.

lpw25 commented:
This latest version is making a debug assertion fail in gc_ctrl.c. It is complaining that there is a white block which has already been swept (i.e. it is behind caml_gc_seep_hp) and which points to a blue block. IIUC the new code this is basically expected -- the "remnants" will now look like this thanks to the `left`, `right`, `prev` and `next` pointers stored in them. I guess that either this check should be relaxed or those pointers should be cleared when making a remnant.

lpw25 commented:
Another option would be to give the remnants `Abstract_tag`

stedolan commented on runtime/freelist.c:
```diff
@@ -1308,40 +1308,46 @@ static int bf_is_in_tree (large_free_block *b)
-    bf_small_fl[wosz].merge = &Next_small (v);
+  CAMLassert (Color_val (v) == Caml_white);
+  CAMLassert (wosz <= BF_NUM_SMALL);
+  if (wosz != 0 && (char *) Hp_val (v) < (char *) caml_gc_sweep_hp){
```
I think we need to also check whether `caml_gc_phase == Phase_sweep` here.

damiendoligez replied:
That's right. Fixed.

stedolan commented:
I like the approach in the new patch. Moving the test to `bf_insert_remnant*` is simpler than testing in `bf_split*` and passing the boolean around. Apart from the possible issue with Phase_sweep, the new patch looks good to me.

To fix the incompatibility with the assertions in gc_ctrl.c (that @lpw25 mentions just above), we could use another colouring scheme. The headers of free blocks in the latest patch here are:

  - Large blocks on free list: `Make_header(len, Caml_blue, 0)`
  - Small blocks on free list (ordered): `Make_header(len, Caml_blue, 0)`
  - Small remnants on free list (unordered): `Make_header(len, Caml_white, 0)`
  - Small remnants not on free list (for sweeping): `Make_header(len, Caml_white, 0)`

We need to distinguish the ordered from the unordered small blocks so that the unordered ones can be popped off during `bf_init_merge`. This is the only reason we need to distinguish them (although previously, we needed to distinguish them so that the sweeper behaved correctly).

So, we could distinguish them by tag rather than by colour, giving the following:

  - Large blocks on free list: `Make_header(len, Caml_blue, 0)`
  - Small blocks on free list (ordered): `Make_header(len, Caml_blue, 0)`
  - Small remnants on free list (unordered): `Make_header(len, Caml_blue, 1)`
  - Small remnants not on free list (for sweeping): `Make_header(len, Caml_white, 0)`

This would mean that the colour of a small remnant varies depending on whether it's added to the freelist, but we could handle this by updating the header in bf_insert_remnant_small.

The invariants of this scheme are a bit simpler: a block is blue iff it's on the freelist. (I think this should be compatible with the consistency checks in `gc_ctrl.c` as well, but haven't tested)

lpw25 commented:
@stedolan Note that your new scheme also requires changing the colour of the blocks at the start of sweeping. Which is probably fine since we are going to read (although not write) to those headers at the start of sweeping anyway.

I can't decide whether the invariant is simpler in your version or not. In one `blue` means "is on the free list" whereas in the other `blue` means "can be efficiently removed from the free list". The second of those is the property that the sweeper actually relies on, which is why your code needs to change some colours at the start of sweeping. Of course the sweeper also relies on only those sorts of nodes being on the freelist, which is why we have to take the others off at the start of sweeping.

damiendoligez commented:
>Another option would be to give the remnants `Abstract_tag`

That was the case previously, and my latest patch removed that because I thought it was needed only for the black blocks. My mistake.

damiendoligez committed:
fix more bugs

damiendoligez commented:
@stedolan simple invariants are nice but I'd rather save the memory accesses.

stedolan commented:
> @stedolan simple invariants are nice but I'd rather save the memory accesses.

Fair enough!

gasche commented:
Meta-note: I think it would be better for the long-term maintenance of the runtime system if we could get some extra pairs of eyes on this codebase before it was merged upstream (because it is hard to motivate people to look at it -- and thus decrease its bus factor -- after it is already there).

We can be careful not to slip past the 4.10 freeze date; I understand having this in 4.10 is important for Jane Street, and (to me personally) it seems entirely reasonable given the solid interactions and reviews for a policy that will not be enabled by default at first.

I coerced^Wconvinced Damien to give a talk about the design of the allocator to a select audiences of potential reviewer-grade attendees (eg. @jhjourdan); the talk was initially planned for yesterday but was delayed to next week. I'm hoping this will give us a last occasion to look at the code before it is merged.

alainfrisch commented:
If the slides of that presentation could be shared, this would be great!  (This is certainly obvious for people actively involved in this PR, but I haven't seen a high-level description of what the new policy actually is, and as a user I would like to understand it.)

gasche commented:
Of course, this is one of the indirect goals of asking for a presentation of the work.

jhjourdan commented on runtime/gc_ctrl.c:
```diff
@@ -493,15 +488,30 @@ CAMLprim value caml_gc_set(value v)
+    caml_compact_heap (newpolicy);
+    caml_gc_message (0x20, "New allocation policy: %"
+                     ARCH_INTNAT_PRINTF_FORMAT "u\n", newpolicy);
+    caml_final_do_calls ();
```
Note that since #8691, we do no longer call `caml_final_do_calls` directly. We instead centralize all the asynchronous callbacks in `caml_check_urgent_gc`. Also, since this can end up in an exception, it would be preferable to do that at the end of the function (after setting the minor heap size).

damiendoligez replied:
Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -31,82 +32,112 @@
     prev = cur;
-    cur = Next (prev);
+    cur = Next_small (prev);
   }
```
I am right that the `prev` variable is never used? We could do the same loop using `cur` only.

damiendoligez replied:
Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -120,78 +151,35 @@ static void fl_check (void)
 */
-static header_t *allocate_block (mlsize_t wh_sz, int flpi, value prev,
-                                 value cur)
+static header_t *nf_allocate_block (mlsize_t wh_sz, int flpi, value prev,
```
Parameter `flpi` is never used.

damiendoligez replied:
Fixed.

bluddy commented:
I'd really appreciate that presentation being recorded. It would serve as a wonderful resource for the community.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -597,27 +1755,147 @@ void caml_make_free_blocks (value *p, mlsize_t size, int do_merge, int color)
+header_t *caml_fl_allocate (mlsize_t wo_sz)
+{
+  return (*p_allocate) (wo_sz);
+}
```
Why not directly exporting the function pointers instead of wrapping the indirect calls in directly-called functions?

damiendoligez replied:
I wanted to make minimal changes to the API of freelist.c. I had to change `merge_block` slightly, but that's unavoidable. I've now exported the direct-call wrappers as inline functions in `freelist.h`.

gasche commented:
Unfortunately it won't: I don't have equipment to record it, and I don't want to put more pressure on the speaker. (This is getting off-topic, I won't comment more about the presentation here.)

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
     while (cur != Val_NULL && Bp_val (cur) < Bp_val (bp)){
-      CAMLassert (Bp_val (prev) < Bp_val (bp) || prev == Fl_head);
+      CAMLassert (Bp_val (prev) < Bp_val (bp) || prev == Nf_head);
       /* XXX TODO: extend flp on the fly */
```
That comment is specific to first-fit, right?

damiendoligez replied:
Removed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    if (cur == caml_fl_merge) merge_found = 1;
+    prev = cur;
+    cur = Next_small (prev);
+  }
```
Idem: prev is not used in this loop.

damiendoligez replied:
Fixed.

xavierleroy commented:
The next thing you know, @damiendoligez will have to write a paper about his new allocator!  :-)

damiendoligez commented:
I'll make the slides available after the talk.

lpw25 commented on runtime/freelist.c:
```diff
@@ -597,27 +1755,147 @@ void caml_make_free_blocks (value *p, mlsize_t size, int do_merge, int color)
+  policy_best_fit = 2,
+} caml_policy_t;
+
+caml_policy_t caml_allocation_policy = policy_next_fit;
```
This change from `uintnat` to `caml_policy_t` needs to be reflected in `gc_ctrl.c` -- at the moment `caml_gc_get` reads the allocation policy and 7 other bytes of memory to return to the user.

xavierleroy replied:
Color me surprised.  I assume this is the relevant line of `caml_gc_get`?
```
  Store_field (res, 6, Val_long (caml_allocation_policy));              /* a */
```
The friendly C compiler will read `sizeof(caml_policy_t)` bytes of memory, then convert the resulting value to an OCaml integer via `Val_long`.  What could possibly go wrong?

lpw25 replied:
I think it is the:
```
extern uintnat caml_allocation_policy
```
line further up in `gc_ctrl.c` that is causing problems.
I defer to you on the vagaries of C compilers and I haven't checked that changing this line fixes it, but the behaviour is definitely consistent with reading the wrong size from memory.

xavierleroy replied:
Ouch!  This is awful indeed.  Thanks for pointing it out.

damiendoligez replied:
Sigh. That's what I get for trying to clean up the code... Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+
+/* Auxiliary functions for bitmap */
+
+/* Find first bit set in a word. */
```
Since "first bit set" is rather ambiguous (it may be interpreted as highest-level bit, since we usually write higher bits first), I would prefer 'lower-level bit set".

damiendoligez replied:
Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+static struct {
+  value free;
+  value *merge;
+} bf_small_fl [BF_NUM_SMALL + 1];
```
Perhaps add a comment to warn that the first cell of the array is wasted?

damiendoligez replied:
Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+  return total_size;
+}
+
+static void bf_check (void)
```
I guess the idea is to use the `xxx_check` functions as a kind of documentation for the invariants of the data structure. If so, would it be possible to ad something to make clearer the invariants on the relative positions of the `bf_small_fl[i].merge` fields, of `caml_fl_merge` and of `caml_gc_sweep_hp`?

damiendoligez replied:
It's more of a debugging aid, but it's true that `check`'s goal is to implement the invariants that are relatively cheap to compute.

Added.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    if (bf_small_fl[i].merge == &bf_small_fl[i].free) merge_found = 1;
+    if (bf_small_fl[i].free != Val_NULL) map |= 1 << (i-1);
+    for (b = bf_small_fl[i].free; b != Val_NULL; b = Next_small (b)){
+      if (bf_small_fl[i].merge == &Next_small (b)) merge_found = 1;
```
If I understand well, the `merge` pointer cannot be in the "white zone" of the free lists. Can this be made clearer in `bf_check`?

damiendoligez replied:
It just means that it points to a pointer that is either NULL or to a blue block.

Added.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+      x = y;
+    }
+  }
+  /* reassemble the tree */
```
I would add `CAMLassert(x->left == NULL)` here to make it easier to compare the code with `bf_splay`.

damiendoligez replied:
Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    r = x->right;
+    INSTR_alloc_jump (1);
+    r->left = l;
+    *p = r;
```
We don't care about the final value of `x->right` here. So we have a simpler version of this code:
```C
bf_splay_least(&r);
r->left = l;
*p = r;
```

damiendoligez replied:
Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+   existing list.
+   Splay if the list is already present.
+*/
+static void bf_insert_block (large_free_block *n)
```
Why not splaying when the list is not present? As far as I understand the amortized complexity analysis for splay trees, splaying is necessary as soon as we access a node, whether it be for reading, modifying, or removing a node.

I think it would even be a good idea to make simplify `bf_search` much by make it splay first and then return either the root or a reference to one of its children, depending on whether the target exists or not.

damiendoligez replied:
Postponed to 4.11, see #9029.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+   and white (or a 0-size fragment).
+   The block may be left out of the list depending on the sweeper's state.
+   The free list size is updated accordingly.
+*/
```
Perhaps comment a bit on why we may need to leave the block out of the freelist.

damiendoligez replied:
Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    CAMLassert (b->prev->next == b);
+    CAMLassert (b->next->prev == b);
+    if (b->isnode){
+      large_free_block **p = bf_search (bf_large_wosize (b));
```
As for insertion, we should splay here too.

damiendoligez replied:
Postponed to 4.11, see #9029.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+   the block is (one of the) smallest in the tree.
+   In this case, the large block becomes (or remains) the smallest
+   in the tree and we set the [bf_large_least] pointer.
+*/
```
This comment is misleading, because it let us think that if `set_least == 1`, then `*p` may not be the smallest block in the tree. Perhaps be more precise by writting "has the smallest size" instead of "one of the smallest".

Also "becomes the smallest" is misleading.

damiendoligez replied:
Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+  if (wosz <= BF_NUM_SMALL){
+    if (bf_small_fl[wosz].free != Val_NULL){
+      /* fast path: allocate from the corresponding free list */
+      block = bf_small_fl[wosz].free;
```
Nitpick (one more) : To minimize memory accesses and make code shorter (and arguably clearer), we could use `block` instead of `bf_small_fl[wosz].free` everywhere below.

damiendoligez replied:
Not quite everywhere, but yes. Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+      return Hp_val (block);
+    }else{
+      /* allocate from the next available size */
+      mlsize_t s = ffs (bf_small_map & ~ ((1 << wosz) - 1));
```
We have: `~ ((1 << wosz) - 1) == ((-1) << wosz)`. So `((-1) << wosz)` is a shorter, and arguably easier to read version of this bit twiddling.

damiendoligez replied:
That's nice. Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+  /* Find the starting point of the current run of free blocks. */
+  if (caml_fl_merge != Val_NULL && Next_in_mem (caml_fl_merge) == bp){
+    start = caml_fl_merge;
+    CAMLassert (Color_val (start) == Caml_blue);
```
How do we know that the assertion below is true? In other words, isn't it possible that `caml_fl_merge` has been allocated (and thus set to `Caml_black`) since we last set `caml_fl_merge`? There is some code in the next-fit and best-fit allocators to prevent this eventuality, but I don't see anything like this here.

damiendoligez replied:
This is resolved by @stedolan's patch (2226818) see below for discussion.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+      caml_fl_cur_wsz += Whsize_wosize (wosz);
+      bf_insert_block ((large_free_block *) bp);
+    }else{
+      Hd_val (bp) = Whitehd_hd (Hd_val (bp));
```
Shouldn't this block be set to `Abstract_tag`?

damiendoligez replied:
Good catch. Note that it would be very hard to trigger this bug: you need a 32-bit *debug* runtime, a heap extension by a size 2 to 32 modulo Max_whsize, a stray pointer inside the newly-allocated block, and a heap check to trigger the assertion.
Fixed.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -597,27 +1755,147 @@ void caml_make_free_blocks (value *p, mlsize_t size, int do_merge, int color)
-      Make_header (Wosize_whsize (sz), 0, color);
-    if (do_merge) caml_fl_merge_block (Val_hp (p));
+    if (do_merge){
+      mlsize_t wosz = Wosize_whsize (sz);
```
Nitpick: we could move this variable initialization higher in the loop to use `wosz` in the `else` branch too.

damiendoligez replied:
Done.

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    caml_fl_cur_wsz += Whsize_val (cur);
+  next:
+    cur = Next_in_mem (cur);
+    if ((char *) cur > limit) goto end_of_run;
```
Isn't it `cur >= limit` ?

lpw25 replied:
cur points to the value rather than the header, so I think `cur > limit` is correct. Iff we have a size zero remnant at the end of the heap (I don't know whether that is actually possible) then we would see `cur = limit` in which case we should continue to merge it onto the current block.

jhjourdan replied:
Ah, you are right! God, this is misleading!

jhjourdan replied:
Perhaps we could write:
```suggestion
    if (cur >= Val_hp(limit)) goto end_of_run;
```
to make this clearer?

damiendoligez replied:
I changed it to `Hp_val(cur) >= limit` to make it homogeneous with line 558 of major_gc.c.
(and less of an undefined behavior)

jhjourdan commented on runtime/freelist.c:
```diff
@@ -31,82 +32,112 @@
-#define policy caml_allocation_policy
+static value nf_prev = Nf_head;  /* Current allocation pointer. */
+static value nf_last = Val_NULL; /* Last block in the list.  Only valid
+                                  just after [caml_fl_allocate] returns NULL. */
```
```suggestion
                                  just after [nf_allocate] returns NULL. */
```

jhjourdan commented on runtime/freelist.c:
```diff
@@ -120,78 +151,35 @@ static void fl_check (void)
 #endif
       /* In case 1, the following creates the empty block correctly.
          In case 0, it gives an invalid header to the block.  The function
          calling [caml_fl_allocate] will overwrite it. */
```
```suggestion
          calling [nf_allocate] will overwrite it. */
```

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+
+#define Ff_head (Val_bp (&(ff_sentinel.first_field)))
+static value ff_last = Val_NULL; /* Last block in the list.  Only valid
+                                  just after [caml_fl_allocate] returns NULL. */
```
```suggestion
                                  just after [ff_allocate] returns NULL. */
```

jhjourdan commented on runtime/freelist.c:
```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+#endif
+      /* In case 1, the following creates the empty block correctly.
+         In case 0, it gives an invalid header to the block.  The function
+         calling [caml_fl_allocate] will overwrite it. */
```
```suggestion
         calling [ff_allocate] will overwrite it. */
```

jhjourdan reviewed (changes requested):
Overall, this looks good. On the cosmetic side, I am not very happy about the code duplication between the next-fit and the first-fit implementations.

My comments are mostly nitpicks and documentation requests, but I have a few more serious concerns:
- In order to satisfy the theoretical guarantees provided by splay trees, we have to splay at every single access, including insertions and removals. Basically, each time we follow a left or right edge, we have to consider splaying it.
- `caml_fl_merge` is not maintained when a block is allocated. As a result, it seems to me that it may not point to a blue block when used in `bf_merge_block`.
- In `bf_merge_block` again, I think there is a typo when checking `cur > limit` I think this should be `cur >= limit`.
- In `bf_add_blocks`, we don't use `Abstract_tag` when inserting the new block.
- Since #8691, we do no longer call `caml_final_do_calls` directly.

jhjourdan commented:
Another remark: some C bindings (e.g., Lablgtk) depend on disabling the compactor because they have trouble registering all their roots. I guess this would mean that they are incompatible with best-fit, because changing the policy requires a compaction.

How should we document that?

lpw25 commented:
> caml_fl_merge is not maintained when a bloc is allocated. As a result, it seems to me that it may not point to a blue block when used in bf_merge_block.

As it happens we've had a couple of segfaults that appear to be an instance of this. I think that the following patch should be sufficient to fix it:
```
diff --git a/runtime/freelist.c b/runtime/freelist.c
index 6ddc44db37..a785d0700a 100644
--- a/runtime/freelist.c
+++ b/runtime/freelist.c
@@ -1483,6 +1486,7 @@ static header_t *bf_alloc_from_large (mlsize_t wosz, large_free_block **p,
 
   CAMLassert (bf_large_wosize (n) >= wosz);
   if (n->next == n){
+    if (caml_fl_merge == (value) n) caml_fl_merge = Val_NULL;
     if (wosize_n > bound + Whsize_wosize (wosz)){
       /* TODO splay at [n]? if the remnant is larger than [wosz]? */
       if (set_least){
@@ -1509,6 +1513,7 @@ static header_t *bf_alloc_from_large (mlsize_t wosz, large_free_block **p,
     CAMLassert (bf_large_wosize (b) == bf_large_wosize (n));
     n->next = b->next;
     b->next->prev = n;
+    if (caml_fl_merge == (value) b) caml_fl_merge = Val_NULL;
     if (wosize_n == wosz){
       caml_fl_cur_wsz -= Whsize_wosize (wosz);
       return Hp_val ((value) b);
@@ -1562,6 +1567,7 @@ static header_t *bf_allocate (mlsize_t wosz)
       if (bf_small_fl[wosz].merge == &Next_small (bf_small_fl[wosz].free)){
         bf_small_fl[wosz].merge = &bf_small_fl[wosz].free;
       }
+      if (caml_fl_merge == block) caml_fl_merge = Val_NULL;
       bf_small_fl[wosz].free = Next_small (bf_small_fl[wosz].free);
       if (bf_small_fl[wosz].free == Val_NULL) unset_map (wosz);
       caml_fl_cur_wsz -= Whsize_wosize (wosz);
@@ -1577,6 +1583,7 @@ static header_t *bf_allocate (mlsize_t wosz)
         if (bf_small_fl[s].merge == &Next_small (bf_small_fl[s].free)){
           bf_small_fl[s].merge = &bf_small_fl[s].free;
         }
+        if (caml_fl_merge == block) caml_fl_merge = Val_NULL;
         bf_small_fl[s].free = Next_small (bf_small_fl[s].free);
         if (bf_small_fl[s].free == Val_NULL) unset_map (s);
         result = bf_split_small (wosz, block);
@@ -1589,6 +1596,7 @@ static header_t *bf_allocate (mlsize_t wosz)
     if (bf_large_least != NULL){
       mlsize_t least_wosz = bf_large_wosize (bf_large_least);
       if (least_wosz > BF_NUM_SMALL + Whsize_wosize (wosz)){
+        if (caml_fl_merge == (value) bf_large_least) caml_fl_merge = Val_NULL;
         result = bf_split (wosz, (value) bf_large_least);
         caml_fl_cur_wsz += Whsize_wosize (least_wosz) - Whsize_wosize (wosz);
           /* remnant stays in tree */
```

jhjourdan commented:
> As it happens we've had a couple of segfaults that appear to be an instance of this. I think that the following patch should be sufficient to fix it:

Alternatively, we can *check* when coalescing that the block pointed by `caml_fl_merge` is still blue.

gasche commented:
The [org-mode slides](http://gallium.inria.fr/~scherer/doc/chameau-sur-le-plateau/2019-10-08-damien-doligez-major-allocator.org) of Damien's talk are now available, as well as the [live notes](http://gallium.inria.fr/~scherer/doc/chameau-sur-le-plateau/2019-10-08-damien-doligez-talk-notes.txt) I took during the talk.

stedolan commented:
> Alternatively, we can check when coalescing that the block pointed by caml_fl_merge is still blue.

Good idea! Here's an alternative patch along those lines:

```
diff --git a/runtime/freelist.c b/runtime/freelist.c
index 6ddc44db37..26f4fc8644 100644
--- a/runtime/freelist.c
+++ b/runtime/freelist.c
@@ -1669,9 +1669,9 @@ static header_t *bf_merge_block (value bp, char *limit)
 
   CAMLassert (Color_val (bp) == Caml_white);
   /* Find the starting point of the current run of free blocks. */
-  if (caml_fl_merge != Val_NULL && Next_in_mem (caml_fl_merge) == bp){
+  if (caml_fl_merge != Val_NULL && Next_in_mem (caml_fl_merge) == bp &&
+      Color_val(caml_fl_merge) == Caml_blue){
     start = caml_fl_merge;
-    CAMLassert (Color_val (start) == Caml_blue);
     bf_remove (start);
   }else{
     start = bp;
```

damiendoligez committed:
fix bug reported by @lpw25 in https://github.com/ocaml/ocaml/pull/8809#pullrequestreview-297585979

damiendoligez commented:
I like this version, it seems more robust and probably less run-time overhead. It relies on the property that `bf_merge_block` is the only function that can turn a header into a non-header.
(and it will never do that to the header under `caml_fl_merge`)

damiendoligez commented:
> In order to satisfy the theoretical guarantees provided by splay trees, we have to splay at every single access, including insertions and removals. Basically, each time we follow a left or right edge, we have to consider splaying it.

A few hours discussing this with @jhjourdan yielded a way of doing that without re-splaying, for all the needed tree functions. It will be a rather large change for a better worst-case guarantee but probably no noticeable speed gain. We've decided to postpone it to 4.11.

damiendoligez committed:
@stedolan's patch from https://github.com/ocaml/ocaml/pull/8809#issuecomment-539539163

xavierleroy commented on runtime/freelist.c:
```diff
@@ -1780,7 +1780,7 @@ typedef enum caml_policy_t {
   policy_best_fit = 2,
 } caml_policy_t;
 
```
Since you're not using the types `caml_policy_t` and `enum caml_policy_t` anywhere else, you could just do
```
enum {
  policy_next_fit = 0,
  policy_first_fit = 1,
  policy_best_fit = 2,
};
```
or, for a more self-documenting style, name the enum but without a typedef:
```
enum caml_policy {
  policy_next_fit = 0,
  policy_first_fit = 1,
  policy_best_fit = 2,
};
```

damiendoligez replied:
I think the `policy_` prefix on each value is self-documenting enough. Fixed.

damiendoligez committed:
fix bug reported by @jhjourdan

damiendoligez committed:
avoid calling caml_final_do_calls directly in caml_gc_set

damiendoligez committed:
remove unused variables, parameter, and comment

damiendoligez committed:
remove unused typedef and enum name

damiendoligez committed:
inline the convenience functions from freelist.c

damiendoligez committed:
add/fix comments

damiendoligez committed:
change wrong call to caml_fl_init_merge in ff_reset

damiendoligez committed:
Small changes suggested by @jhjourdan

damiendoligez commented:
>Another remark: some C bindings (e.g., Lablgtk) depend on disabling the compactor because they have trouble registering all their roots. I guess this would mean that they are incompatible with best-fit, because changing the policy requires a compaction.

FTR: you can choose the initial policy with the OCAMLRUNPARAM environment variable.

damiendoligez committed:
add some invariants to bf_check

damiendoligez committed:
update Changes

gasche commented:
Congratulation on everyone involved, this is a big, nice change. We have to make sure in the release process that heavy users know to try this new policy, and that they do come and make reports to us about improvements and, more importantly, any potential regression.
